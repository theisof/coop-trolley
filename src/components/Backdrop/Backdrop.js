import React, { Component } from 'react'

window.openCoopBackdrops = 0

class Backdrop extends Component {
  constructor(props) {
    super(props);

    this.state = {
      initialLoad: false
    }

    this.scrollPosition = 0
    this.lockDocument = this.lockDocument.bind(this)
    this.releaseDocument = this.releaseDocument.bind(this)
  }

  componentDidMount() {
    this.setState({ initialLoad: true })
  }

  shouldComponentUpdate (nextProps) {
    return this.props.visible !== nextProps.visible
  }

  componentDidUpdate() {
    const { visible } = this.props

    if (visible) {
      this.lockDocument()
    } else {
      this.releaseDocument()
    }
  }

  lockDocument() {
    // Note that execution order is important to get the right dom-behaviour
    this.scrollPosition = window.pageYOffset
    document.body.style.paddingRight = `${this.getScrollBarWidth()}px`
    document.body.classList.add("coop-backdrop-body--overlay-open");
    document.body.style.top = `-${this.scrollPosition}px`
    window.openCoopBackdrops = window.openCoopBackdrops + 1
  }

  releaseDocument() {
    if (window.openCoopBackdrops === 1) {
      document.body.classList.remove("coop-backdrop-body--overlay-open");
      window.scrollTo(0, this.scrollPosition)
      document.body.style.top = `0px`
      document.body.style.paddingRight = `0px`
    }

    window.openCoopBackdrops = window.openCoopBackdrops - 1
  }

  getScrollBarWidth() {
    return window.innerWidth - document.body.clientWidth
  }

  render () {
    const { visible, onClose, zIndex = 100000 } = this.props
    const invisibleClass = (! visible && this.state.initialLoad) ? 'coop-backdrop--invisible' : ''
    const visibilityClass = visible ? 'coop-backdrop--visible' : invisibleClass

    return (
      <div
        style={{ zIndex }}
        className={`coop-backdrop ${visibilityClass}`}
        onClick={onClose}
      ></div>
    )
  }
}

export default Backdrop
