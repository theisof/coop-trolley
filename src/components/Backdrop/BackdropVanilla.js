class BackdropVanilla {
  constructor() {
    this.namespace = 'coop-backdrop-vanilla'
    this.scrollPosition = 0
    this.id = Math.floor((Math.random() * 10000000) + 1)
  }

  open() {
    this.insertStyles()

    this.scrollPosition = window.pageYOffset

    document.body.style.paddingRight = `${this.getScrollBarWidth()}px`
    document.body.classList.add(`${this.namespace}--overlay-open`)
    document.body.style.top = `-${this.scrollPosition}px`
    document.getElementById(this.id).classList.remove('coop-backdrop-vanilla--invisible')
    document.getElementById(this.id).classList.add('coop-backdrop-vanilla--visible')

    window.openCoopBackdrops = window.openCoopBackdrops + 1
  }

  close() {
    if (window.openCoopBackdrops === 1) {
      document.body.classList.remove(`${this.namespace}--overlay-open`);
      window.scrollTo(0, this.scrollPosition)
      document.body.style.top = `0px`
      document.body.style.paddingRight = `0px`
    }

    document.getElementById(this.id).classList.remove('coop-backdrop-vanilla--visible')
    document.getElementById(this.id).classList.add('coop-backdrop-vanilla--invisible')

    window.openCoopBackdrops = window.openCoopBackdrops - 1
  }

  getScrollBarWidth() {
    return window.innerWidth - document.body.clientWidth
  }

  insertStyles() {
    const id = `${this.namespace}__styles`

    if (! document.getElementById(id)) {
      let tag = document.createElement('style');

      tag.id = id
      tag.innerHTML = styles

      document.body.insertBefore (tag, document.body.firstChild)
    }
  }

  init() {
    window.openCoopBackdrops = window.openCoopBackdrops || 0

    let rootElem  = document.createElement("div")

    rootElem.className = this.namespace
    rootElem.id = this.id
    document.body.insertBefore (rootElem, document.body.firstChild)

    const element = document.getElementById(this.id)

    element.addEventListener('click', this.close)
  }
}


const styles = `
  .coop-backdrop-vanilla {
    position: fixed;
    width: 100%;
    height: 100vh !important;
    overflow: hidden;
    background-color: rgba(0,0,0,0.5);
    left: 0;
    top: 0;
    visibility: hidden;
    opacity: 0;
  }

  .coop-backdrop-vanilla--visible {
    animation: fadein-backdrop 0.5s forwards;
  }

  .coop-backdrop-vanilla--invisible {
    animation: fadeout-backdrop 0.5s forwards;
  }

  body.coop-backdrop-vanilla--overlay-open {
    overflow: hidden !important;
    position: fixed !important;
    width: 100% !important;
    height: 100% !important;
  }

  @keyframes fadein-backdrop {
    from {
      visibility: hidden;
      opacity: 0;
    }
    to   {
      visibility: visible;
      opacity: 1;
    }
  }

  @keyframes fadeout-backdrop {
    from {
      visibility: visible;
      opacity: 1;
    }
    to   {
      visibility: hidden;
      opacity: 0;
    }
  }
`

export default BackdropVanilla
