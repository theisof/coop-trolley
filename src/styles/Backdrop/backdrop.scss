.coop-backdrop {
  position: fixed;
  width: 100%;
  height: 100vh !important;
  overflow: hidden;
  background-color: rgba(0,0,0,0.5);
  left: 0;
  top: 0;
  visibility: hidden;
  opacity: 0;
}

.coop-backdrop--visible {
  animation: fadein-backdrop 0.5s forwards;
}

.coop-backdrop--invisible {
  animation: fadeout-backdrop 0.5s forwards;
}

body.coop-backdrop-body--overlay-open {
  overflow: hidden !important;
  position: fixed !important;
  width: 100% !important;
  height: 100% !important;
}

@keyframes fadein-backdrop {
  from {
    visibility: hidden;
    opacity: 0;
  }
  to   {
    visibility: visible;
    opacity: 1;
  }
}

@keyframes fadeout-backdrop {
  from {
    visibility: visible;
    opacity: 1;
  }
  to   {
    visibility: hidden;
    opacity: 0;
  }
}