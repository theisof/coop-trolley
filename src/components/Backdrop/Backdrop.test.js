import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json';
import Backdrop from './Backdrop'

it('Backdrop renders correctly', () => {
  const tree = shallow( <Backdrop />)

  expect(toJSON(tree)).toMatchSnapshot()
})
