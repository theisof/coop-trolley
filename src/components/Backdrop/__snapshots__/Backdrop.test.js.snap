// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop renders correctly 1`] = `
<div
  className="coop-backdrop "
  style={
    Object {
      "zIndex": 100000,
    }
  }
/>
`;
